### PR TITLE
Introduce feature allowing modifying parent/child relationships.

### LINKED ISSUE #25: Allow pages to be moved to be children of other pages
Doesn't move pages to child?

### PR DESCRIPTION
**Pull request by peterwilsoncc**
### Description of the Change

Adds row actions to move a page up or down the page hierarchy.

The "Move Out from under Parent Page name" action:
* modifies the parent of child pages to their grandparent
* pages without a grandparent are set to top level pages
* the child tree of the post is unmodified, it moves with the parent.

The "Move Under Page Name" action
* sets top level pages' parent to the post immediately above it in the list
* sets existing child pages to be the child of the preceding child post on the same level
* the child tree of the post is unmodified, it moves with the parent

At the moment this always triggers a full page refresh. 

Closes #25.

### How to test the Change

1. Generate some pages with `wp post generate --post_type=page`
  * these pages will not have a menu_order specified
  * the tests below should work as expected without the menu_order
3. Click the "Move Under Page Name" action of a page
4. The post should be made a child of the preceding page in the list
5. Click "Move Under Page Name" of the page following the page you've already edited.
6. The second page should be made a child of the same parent
7. Attempt to move the first page in the list in or out (you'll need to get the action to display via the CSS browser tools), ensure nothing happens to the posts parent setting.
8. Click "Move Under Page Name" of the final child page in your tree
9. This should be made a child of the page above it, a grandchild of the previous parent
10. Click "Move Under Page Name" of the first child page
11. The entire tree should be move in one. The page you clicked should become a top level page

### Changelog Entry

> Added - Ability to modify the page hierarchy.

### Credits

Props @peterwilsoncc.

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [x] All new and existing tests pass.

### CODE REVIEW

File: class-simple-page-ordering.php [hunk 1]
```diff
@@ -199,6 +358,53 @@ public static function admin_head() {
...
+				esc_attr( sprintf( __( 'Move &#8220;%s&#8221; in', 'simple-page-ordering' ), $title ) ),
+				__( '< Move in', 'simple-page-ordering' )
+			);
+
+			$actions['spo-move-out'] = sprintf(
+				'<a href="%s" rel="bookmark" aria-label="%s">%s</a>',
+				esc_url( $move_out_link ),
+				/* translators: %s: Post title. */
+				esc_attr( sprintf( __( 'Move &#8220;%s&#8221; out', 'simple-page-ordering' ), $title ) ),
+				__( 'Move out >', 'simple-page-ordering' )
```
Comment (peterwilsoncc) [inline class-simple-page-ordering.php]: I'm not super keen on move in / move out as the text links -- it needs something clearer but also short so it fits the actions toolbar. Suggestions are encouraged, thanks.

**Comment by peterwilsoncc** [comment]
Some shell commands for getting a bunch of posts with the same ID in the title as the post ID.

Warning: it empties the entire post table. 


```sh
wp site empty
wp post generate --post_type=page
for i in {1..100}; do wp post update $i --post_title="Page $i"; done;

# To set menu order to the ID too
for i in {1..100}; do wp post update $i --post_title="Page $i" --menu_order=$i; done;
```

**Comment by shannonmfisher** [comment]
Hello hi! 

After some noodling, I think the best option is `Nest page` and `Un-nest page` (could be nest/un-nest if the page part make it too long, but I'd leave the page part if possible to be super clear). 

`Make subpage` is another option, but I think it's too complicated because there's no succinct opposite and because nesting it further isn't as intuitive with this command. 

`Rank` page higher/lower could be another option, but I don't think "rank" is intuitive and it's long. 

`Order` or `level` page up/down are also possibilities, but again, not as intuitive as `nest`. 

Lmk if any of these feel right or if you'd like me to keep thinking on it!

**Comment by ankitguptaindia** [comment]
Hello @peterwilsoncc, I have just tested this new edition, and it works well for me. I was wondering if we could implement a similar behavior for moving in and out of pages by using drag and drop? similar to how we navigate up and down pages.

I am not familiar with the technical aspects of this change, but I wanted to share the idea and get your feedback.

I noticed a similar flow in menu ordering as well: 

[Screen recording (1).webm](https://github.com/10up/simple-page-ordering/assets/6384299/46c0a182-71c5-40aa-b370-321824ce2ed4)

**Comment by peterwilsoncc** [comment]
@ankitguptaindia I had an attempt at that but couldn't get it working in an intuitive fashion. As the pages use tables rather than lists, the jQuery UI library only expects items to move up or down, I attempted to fake it but the drop targets' behaviour wasn't great.

**Comment by jeffpaul** [comment]
@peterwilsoncc besides tweaking the copy, is there anything else needed here before this can move forward towards review/release?

**Comment by peterwilsoncc** [comment]
@jeffpaul That should do.

I'm worried `un nest` implies that it moves the page to the top level (ie without a parent) but I can push the change and we can get some real-world feedback and improve if required.

**Comment by jeffpaul** [comment]
@peterwilsoncc I'm fine with you YOLOing the decision on the copy and we can iterate as needed if there's a consensus in the community for something different

**Comment by peterwilsoncc** [comment]
@jeffpaul I've just pushed some changes that alter the copy to match that of the native menu pages in WordPress Core. 

Screen Shot 2024-01-18 at 11 44 51 am

"Move Under Page Name" sets the page as a child of page name
"Move Out from under Parent Page name" sets the page as a child of it's grandparent or top level if no grandparent exists.

I think the copy is clearer than move in/move out and matching existing core text is probably the best thing to do here.

**Review by faisal-alvi — APPROVED** [review]
@peterwilsoncc thank you The PR LGTM and also work as expected.

image

Only the description should be updated that mentions the "Move in"/"Move out" links.

**Comment by peterwilsoncc** [comment]
@faisal-alvi I've updated the description with the new link text.

**Pull request merged**